accounting/src/delta/delta_data_collection/mod.rs

```diff
@@ -101,134 +143,119 @@ impl<K: Ord + Copy, T> DeltaDataCollection<K, T> {
-    fn from_iter<I: IntoIterator<Item = (K, DataDelta<T>)>>(iter: I) -> Self {
-        DeltaDataCollection {
-            data: BTreeMap::<K, DataDelta<T>>::from_iter(iter),
+fn combine_delta_elements<T: Clone + PartialEq>(
+    lhs: &DeltaMapElement<T>,
+    rhs: DeltaMapElement<T>,
+) -> Result<DeltaMapOp<T>, Error> {
+    match (lhs, rhs) {
+        (DeltaMapElement::Delta(d1), DeltaMapElement::Delta(d2)) => {
+            match combine_delta_data(d1, d2)? {
+                Some(d) => Ok(DeltaMapOp::Insert(DeltaMapElement::Delta(d))),
+                None => Ok(DeltaMapOp::Erase),
+            }
+        }
+        (DeltaMapElement::Delta(d), DeltaMapElement::DeltaUndo(u)) => {
+            match combine_delta_data(d, u.0)? {
+                Some(d) => Ok(DeltaMapOp::Insert(DeltaMapElement::Delta(d))),
+                None => Ok(DeltaMapOp::Erase),
+            }
+        }
+        (DeltaMapElement::DeltaUndo(_), DeltaMapElement::Delta(_)) => {
+            Err(Error::DeltaOverUndoApplied)
+        }
+        (DeltaMapElement::DeltaUndo(u1), DeltaMapElement::DeltaUndo(u2)) => {
+            match combine_delta_data(&u1.0, u2.0).map(|d| d.map(|d| DataDeltaUndo(d)))? {
+                Some(d) => Ok(DeltaMapOp::Insert(DeltaMapElement::DeltaUndo(d))),
+                None => Ok(DeltaMapOp::Erase),
+            }
         }
     }
 }
 
 /// Given two deltas, combine them into one delta, this is the basic delta data composability function
-fn combine_delta_data<T>(
+fn combine_delta_data<T: Clone + PartialEq>(
     lhs: &DataDelta<T>,
     rhs: DataDelta<T>,
-) -> Result<DeltaMapOp<DataDelta<T>>, Error> {
+) -> Result<Option<DataDelta<T>>, Error> {
     match (lhs, rhs) {
         (DataDelta::Create(_), DataDelta::Create(_)) => Err(Error::DeltaDataCreatedMultipleTimes),
-        (DataDelta::Create(_), DataDelta::Modify(d)) => Ok(DeltaMapOp::Write(DataDelta::Create(d))),
-        (DataDelta::Create(_), DataDelta::Delete) => {
-            // if lhs had a creation, and we delete, this means nothing is left and there's a net zero to return
-            Ok(DeltaMapOp::Delete)
+        (DataDelta::Create(lhs), DataDelta::Modify(from, to)) => {
+            ensure!(lhs == &from, Error::DeltaDataMismatch);
+            Ok(Some(DataDelta::Create(to)))
+        }
+        (DataDelta::Create(lhs), DataDelta::Delete(rhs)) => {
+            // if lhs had a creation, and we DataDeltaUndo::Delete, this means nothing is left and there's a net zero to return
+            ensure!(lhs == &rhs, Error::DeltaDataMismatch);
+            Ok(None)
+        }
+        (DataDelta::Modify(_, _), DataDelta::Create(_)) => {
+            Err(Error::DeltaDataCreatedMultipleTimes)
+        }
+        (DataDelta::Modify(lhs1, lhs2), DataDelta::Modify(rhs1, rhs2)) => {
+            if lhs1 == &rhs2 && lhs2 == &rhs1 {
+                Ok(None)
+            } else {
+                Ok(Some(DataDelta::Modify(lhs1.clone(), rhs2)))
```

Comment:
Why the clone here? Seems `lhs1` is consumed by the condition, changing it to borrow `&lhs1 == &rhs2` should make it possible to avoid the clone here.

Reply:
In case of create/delete and create/modify pairs, there is an `ensure!` verifying the intermediate values match but not here. That is an inconsistency.